libtransmission/session-alt-speeds.h

```diff
@@ -0,0 +1,158 @@
+
+/** Manages alternate speed limits and a scheduler to auto-toggle them. */
+class tr_session_alt_speeds
+{
+public:
+    enum class ChangeReason
+    {
+        User,
+        Scheduler
+    };
+
+    class Mediator
+    {
+    public:
+        virtual ~Mediator() noexcept = default;
+
+        using ChangeReason = tr_session_alt_speeds::ChangeReason;
+        virtual void isActiveChanged(bool is_active, ChangeReason reason) = 0;
+
+        [[nodiscard]] virtual time_t time() = 0;
+    };
+
+    constexpr explicit tr_session_alt_speeds(Mediator& mediator) noexcept
+        : mediator_{ mediator }
+    {
+    }
+
+    void load(tr_variant* src);
+    void save(tr_variant* tgt) const;
+    static void defaultSettings(tr_variant* tgt);
+
+    [[nodiscard]] constexpr bool isActive() const noexcept
+    {
+        return is_active_;
+    }
+
+    void checkScheduler();
+
+    void setSchedulerEnabled(bool enabled)
+    {
+        scheduler_enabled_ = enabled;
+        updateScheduler();
+    }
+
+    // return true iff the scheduler will turn alt speeds on/off
+    [[nodiscard]] constexpr auto isSchedulerEnabled() const noexcept
+    {
+        return scheduler_enabled_;
+    }
+
+    void setStartMinute(size_t minute)
+    {
+        minute_begin_ = minute;
+        updateScheduler();
+    }
+
+    [[nodiscard]] constexpr auto startMinute() const noexcept
+    {
+        return minute_begin_;
+    }
+
+    void setEndMinute(size_t minute)
+    {
+        minute_end_ = minute;
+        updateScheduler();
+    }
+
+    [[nodiscard]] constexpr auto endMinute() const noexcept
+    {
+        return minute_end_;
+    }
+
+    void setWeekdays(tr_sched_day days)
+    {
+        use_on_these_weekdays_ = days;
+        updateScheduler();
+    }
+
+    [[nodiscard]] constexpr tr_sched_day weekdays() const noexcept
+    {
+        return static_cast<tr_sched_day>(use_on_these_weekdays_);
+    }
+
+    [[nodiscard]] constexpr auto limitKBps(tr_direction dir) const noexcept
+    {
+        return dir == TR_DOWN ? speed_down_kilobytes_per_second_ : speed_up_kilobytes_per_second_;
+    }
+
+    constexpr void setLimitKBps(tr_direction dir, size_t limit) noexcept
+    {
+        if (dir == TR_DOWN)
+        {
+            speed_down_kilobytes_per_second_ = limit;
+        }
+        else
+        {
+            speed_up_kilobytes_per_second_ = limit;
+        }
+    }
+
+    void setActive(bool active, ChangeReason reason);
+
+private:
+    Mediator& mediator_;
+
+    void updateScheduler();
+    void updateMinutes();
+
+    // whether `time` hits in one of the `minutes_` that is true
+    [[nodiscard]] bool isActiveMinute(time_t time) const noexcept;
+
+    static auto constexpr MinutesPerHour = int{ 60 };
+    static auto constexpr MinutesPerDay = int{ MinutesPerHour * 24 };
+    static auto constexpr MinutesPerWeek = int{ MinutesPerDay * 7 };
+
+    // is turtle mode on right now?
```

Comment:
"turtle mode" is being rename, right?

Reply:
Fixed in c2a48523ceeac8ccc9811d03f914337c33a03840